{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "69870875-aa2d-404a-815e-cb41512d0275",
   "metadata": {},
   "source": [
    "# What to expect\n",
    "\n",
    "We will start our analysis from the mapped reads. The mapping was done using the package [STAR](https://pubmed.ncbi.nlm.nih.gov/23104886/). In this notebook we will take a look at the outputs generated by STAR for the <i>Schistosoma mansoni</i> dataset. We will guide you through the process of combining the results for each sample in the dataset into a single dataframe and viewing them using Principal Components Analysis. In the second part of this session, you will repeat this process for your choice of dataset in notebook 2B."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "713ebba8-cd89-4b38-ad93-f10fd7780387",
   "metadata": {},
   "source": [
    "# The files\n",
    "\n",
    "For each dataset, you will find some information in `data/<dataset>`:\n",
    "\n",
    "1. `README` file with information on how the data was generated\n",
    "2. `list_ids` file with the names of the samples - called accession numbers\n",
    "3. `metadata` file with information on the experimental conditions for each sample\n",
    "\n",
    "For each dataset, you will also find some files stored in `analysis/<dataset>`:\n",
    "\n",
    "1. `star` folder - contains some of the outputs generated by STAR (others have been omitted to save space)\n",
    "2. `multiqc` folder - contains a summary of the quality of mappings by STAR\n",
    "\n",
    "We will use these files to complete our analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "523a81db-c082-43bf-8ed5-282840df8afd",
   "metadata": {},
   "source": [
    "# Checking the quality of mapping\n",
    "We have mapped our reads to the genome, so we can compare expression levels between our samples. But before moving on to the next step, we should check the quality of our mapping. \n",
    "\n",
    "As explained before, the number of reads that map to a gene gives us a measure of how much that gene is being expressed. Therefore, for us to be confident in our quantification, the majority of reads should map to one and only one gene in the reference genome. There are valid reasons why a read can map to more than one gene, for example in the case of families of highly related genes. Reasons to be concerned include insufficiently stringent mapping criteria, short reads that can map to several genes non-specifically, or other technical reasons. As a standard guide, a good quality sample will usually have at least 75% of the reads uniquely mapped.\n",
    "\n",
    "One of the tools frequently used to check mapping quality is [MultiQC](https://multiqc.info/). MultiQC analyses the mapping for each of our samples and produces an HTML report summarizing it. We have run MultiQC on our files and provided it in the analysis subdirectories. Try opening `analysis/Schistosoma_mansoni/multiqc/multiqc_report.html` by double-clicking (you may need to click 'Trust HTML' in a banner)."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bdad12aa-9162-4303-bc0c-61e7ed10a6fd",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-warning\">\n",
    "\n",
    "Discuss in your group:\n",
    "\n",
    "- What % of reads mapped to only one gene of the reference genome?\n",
    "- Are there any samples that look less good? In what way? How might this impact your results?"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7200a165-1af4-442e-b879-71b5bc0efd64",
   "metadata": {},
   "source": [
    "You can add some notes from your discussion on this text cell\n",
    "\n",
    "\\\n",
    "\n",
    "\\\n",
    "\n",
    "\\"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ce758727-2672-48aa-973d-e2c09bbf8379",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-success\">\n",
    "<b>Learning outcomes</b>\n",
    "\n",
    "You should now...\n",
    "\n",
    "- Know how to interpret a MultiQC report\n",
    "- Understand why it is important to check the quality of mapping"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "30a15365-0024-41de-9f80-530606796380",
   "metadata": {},
   "source": [
    "# Combining data across samples\n",
    "\n",
    "Now that we are happy with our mapping, we can move on with the analysis. \n",
    "\n",
    "### Looking at mapping results\n",
    "In the `star` folder there is a file for each of our samples. Each file contains the mapping result for that particular sample. Let's have a look at the mapping results for one of the samples."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "74ee6cfd-e5a7-46d5-b464-093593c53046",
   "metadata": {},
   "outputs": [],
   "source": [
    "! head -n 30 analysis/Schistosoma_mansoni/star/ERR022875ReadsPerGeneUnst.out.tab"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ccc99936-89b8-49df-b272-4d1595b5723e",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "source": [
    "It would be very inconvenient if we had to open each file separately in order to look at the mapping results. Therefore, we will combine all the results in one dataframe. It will be our master dataframe. \n",
    "\n",
    "### Getting the list of samples in python\n",
    "\n",
    "To create the master dataframe, we first need to provide python with a list of samples (accessions). These are stored in the file `data/Schistosoma_mansoni/list_ids.txt`. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9f5b3423-01d9-4fb6-9480-36b845929cd9",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-warning\">\n",
    "Create a list that contains all the accession numbers (sample names) in our experiment.  "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "21cf7d64-945c-4d00-8a3c-97f0c0347522",
   "metadata": {},
   "source": [
    "<details>\n",
    "<summary><i>Hints and tips</i></summary>\n",
    "    \n",
    "    We practised this in the previous notebook. If you need to, go back and use that code.\n",
    "\n",
    "</details>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6b42de56-8d6c-4aec-8564-3d17864109cd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create a new list called accessions. The list will be empty for now. \n",
    "\n",
    "# open the file in read mode\n",
    "# use a for loop over the lines in the file\n",
    "# remove any whitespace/newline characters from the line\n",
    "# add the new accession to the list\n",
    "\n",
    "# have a look at the list to check it all worked well"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "648cba59-81a7-4fd7-b230-24f91bd17b36",
   "metadata": {},
   "source": [
    "### Create the master dataframe\n",
    "Now that we have our accessions list, we are going to use it to create our master dataframe.  \n",
    "\n",
    "We will work with pandas dataframes, as we did in notebook 1.\n",
    "\n",
    "The dataframe should have the gene name as index, and a column of reads per gene for each sample. Each column should have the accession number of that sample as column name.\n",
    "\n",
    "We need some code that:\n",
    "- for each accession number in the accessions list, reads the corresponding `ReadsPerGeneUnst.out.tab` file\n",
    "- Puts the contents of that file into a dataframe that has the gene names as index and a column with the accession number as column name, and the counts for each gene in each row. \n",
    "- Adds that column we just generated to a master dataframe using `join` . You used `join` in notebook 1.\n",
    "\n",
    "Because this is again a repetitive process (the code has to do the same thing for each of our samples) we will need to write another loop."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "45f12d42-36a7-40e2-9288-7acd6c2ef7ca",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-warning\">\n",
    "    \n",
    "Try to fill the gaps `...` in the code below to create the master dataframe."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3d27dacd-a1fc-46fc-bfc0-fdbb3c591a03",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import pandas library\n",
    "import pandas as pd\n",
    "\n",
    "# Create the master dataframe\n",
    "master_df = pd.DataFrame()\n",
    "\n",
    "\n",
    "# Create a loop that for each accession in the list \"accessions\", will:\n",
    "# (1) print the accession number (this allows us to check that the code is working well)\n",
    "# (2) read the ReadsPerGeneUnst.out.tab file for that same accession number into a temporary dataframe \"counts_df\" that has\n",
    "# the gene names column as index and has the column names [\"gene\", \"reads\"]\n",
    "# (3) copy the gene counts data (\"reads\") to a new column named as the accession number\n",
    "# (4) Extract that new column with the current accession's read counts into a smaller dataframe \"accession_df\"\n",
    "# (5) use a dataframe join to add this column to the master dataframe. Note that not all samples might express the same genes, \n",
    "# so the indexes might not be the same. Think about the type of join you need here.\n",
    "\n",
    "for ...:\n",
    "    print(accession)\n",
    "    file = f\"analysis/Schistosoma_mansoni/star/{accession}ReadsPerGeneUnst.out.tab\"\n",
    "    counts_df = ...\n",
    "    ...\n",
    "    ...\n",
    "    ...\n",
    "    master_df = master_df.join(...)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "73bab10a-5256-494e-9ae5-53c4cb7e4411",
   "metadata": {},
   "source": [
    "Let's have a look at our master dataframe we just created"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "01333ca5-1d31-4291-b68b-fe3982b9b6fb",
   "metadata": {},
   "outputs": [],
   "source": [
    "master_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "11cb074a-25f5-4611-b7ff-21738a8c9a53",
   "metadata": {},
   "source": [
    "### Filter and save the master dataframe\n",
    "\n",
    "Hopefully you can see that, although the structure of the dataframe looks good, we need to get rid of the first 5 rows, as they contain summary information instead of gene counts. In DExB2, you already learned ways to either remove rows (week 2, class 3) or to slice the dataframe (week 4, class 8).\n",
    "\n",
    "In addition, it is likely that not all samples express the same genes. If reads were mapped to a gene for sample 1, but not for sample 2, in our dataframe we will have a number of reads for sample 1 and an empty value \"NaN\" for sample 2. We want to replace those empty values with 0. You already practised how to change NaN values in a dataframe in notebook 1. \n",
    "\n",
    "We also want to save the master dataframe as a csv file. \n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "edf42955-75d1-4787-a25d-caa6e9338bf8",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-warning\">\n",
    "\n",
    "Try to fill the gaps `...` in the code below to clean the dataframe and save it as csv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cbac4f75-1670-4b09-9e08-9b8f2314e879",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Keep only the relevant rows\n",
    "master_df_clean = ...\n",
    "# Replace any empty values with 0\n",
    "...\n",
    "# save the dataframe as a csv file called ReadsPerGene\n",
    "...\n",
    "# have a look at the cleaned master_df\n",
    "master_df_clean"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8c2940ec-6f66-4e9f-abbf-30f4db9d48c6",
   "metadata": {},
   "source": [
    "# Normalisation\n",
    "The number of reads mapping to each gene is proportional to the number of expressed transcripts of that gene in that sample. However, there are other factors which affect the number of reads that map to each gene. When comparing gene expression, we need to take those factors into account to minimise their impact, so that our comparisons are reliable. Normalisation is the process of scaling the raw counts to account for these other factors so that the expression levels are more comparable.\n",
    "\n",
    "The main factors that we should consider are:\n",
    "\n",
    "-\tSequencing depth: this is the total number of reads obtained (the “Gene counts” in our MultiQC report). As you saw in the MultiQC report, this was different between samples. \n",
    "\n",
    "-\tGene length: longer genes will usually have more reads than shorter genes.\n",
    "\n",
    "-\tOverall RNA composition: the presence of a small number of very highly expressed genes, or differences in the number of genes expressed between samples can skew our analysis.\n",
    "\n",
    "\n",
    "There are different methods for normalisation, depending on the comparison we want to perform. For example, an analysis might look within a single sample to see which genes are more highly expressed in that sample; or it might compare the expression of a gene between 2 different samples.  \n",
    "\n",
    "The following table is taken from a [Harvard Chan Bioinformatics Core tutorial](https://hbctraining.github.io/DGE_workshop/lessons/02_DGE_count_normalization.html) on the topic:\n",
    "\n",
    "|                                 Normalisation method                                |                                                          Description                                                         |                  Accounted factors                 |                                                                                             |\n",
    "|:-----------------------------------------------------------------------------------:|:----------------------------------------------------------------------------------------------------------------------------:|:--------------------------------------------------:|:-------------------------------------------------------------------------------------------------------------------:|\n",
    "| CPM (counts per million)                                                            | counts scaled by total number of reads                                                                                       | sequencing depth                                   |  |\n",
    "| TPM (transcripts per kilobase million)                                              | counts per length of transcript (kb) per million reads mapped                                                                | sequencing depth and gene length                   |             |\n",
    "| DESeq2’s median of ratios                                                      | counts divided by sample-specific size factors | sequencing depth and RNA composition               |                        |\n",
    "| EdgeR’s trimmed mean of M values (TMM)                                        | uses a weighted trimmed mean of the log expression ratios between samples                                                    | sequencing depth, RNA composition, and gene length |                                                |"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2b54d196-d2ce-4680-97ac-a2c42aec9df4",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-warning\">\n",
    "Discuss in your group which normalisation method might be more appropriate for our analysis."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4462e6b8-d2d5-4f39-ba56-a8f048912cec",
   "metadata": {},
   "source": [
    "You can add some notes from your discussion in this text cell\n",
    "\n",
    "\\\n",
    "\n",
    "\\\n",
    "\n",
    "\\"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1e1c7617-4fd2-467d-952c-00e8a035784a",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-success\">\n",
    "<b>Learning outcomes</b>\n",
    "    \n",
    "You should now know...\n",
    "\n",
    "- What normalisation is \n",
    "- Which factors may need to be controlled for different types of gene count comparisons and why"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d901a274-5990-43d1-80c5-e6856c08197a",
   "metadata": {},
   "source": [
    "In this case, we will use the DESeq2 inbuilt method for normalisation. To implement it, we will use the python package [PyDESeq2](https://pydeseq2.readthedocs.io/en/latest/api/index.html). \n",
    "\n",
    "This method requires two inputs:\n",
    "\n",
    "- a table with all our counts, just like the dataframe we have just created, but it needs to be transposed, so the sample names are in the first column. You already learned how to transpose a dataframe in DExB2, week 2, class 4.\n",
    "- a metadata table that specifies what each sample is"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c84358e6-aadf-46b6-a4ef-3527ab8b3155",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-warning\">\n",
    "    \n",
    "Create the counts and the metadata tables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8dde76c3-545b-4c74-9f75-feab4b396d5e",
   "metadata": {},
   "outputs": [],
   "source": [
    "#create the counts matrix by transposing our master_df\n",
    "counts = \n",
    "\n",
    "#create the metadata table. The index should be the accession number.\n",
    "metadata = \n",
    "\n",
    "#let's have a look at the metadata to make sure it looks right. \n",
    "print(metadata)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "14d04aa7-1454-464d-b17b-8ac8379e0ce4",
   "metadata": {},
   "source": [
    "We are now ready to go ahead with the PyDESeq2 analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "502234a8-9b36-47f7-b3ca-f6cc798c82ac",
   "metadata": {},
   "outputs": [],
   "source": [
    "#To start the analysis, let's intall PyDESeq2\n",
    "! pip install --quiet pydeseq2\n",
    "\n",
    "# and import the \"DeseqDataSet\" class from the dds module of PyDESeq2\n",
    "from pydeseq2.dds import DeseqDataSet"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2ae89cad-695a-4c90-b3da-81b59dc4b473",
   "metadata": {},
   "source": [
    "Now, we will use the DeseqDataSet class to create a \"dds\" object, which is an annotated data matrix, called [AnnData](https://anndata.readthedocs.io/en/latest/generated/anndata.AnnData.html#anndata.AnnData). To create the dds object, we provide some arguments: \n",
    "\n",
    "- `counts` is our transposed master_df matrix\n",
    "- `metadata` is our metadata table that we just created\n",
    "- `refit_cooks` indicates whether Cook's distance should be recalculated during the analysis. You do not have to worry about the details, we will just indicate that we do want to recalculate Cook's distance.\n",
    "- `design` is where we indicate what we want to use to compare samples. In our experiment we want to compare gene expression between the different developmental stages, so we specify \"stage\" as our design factor"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "44f5d4da-c1cc-47e9-820f-fb97acee2074",
   "metadata": {},
   "outputs": [],
   "source": [
    "dds = DeseqDataSet(\n",
    "    counts=counts,\n",
    "    metadata=metadata,\n",
    "    refit_cooks=True,\n",
    "    design=\"stage\"\n",
    ")\n",
    "\n",
    "dds"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c040ea5e-4f97-4b5f-90a8-2667755c921d",
   "metadata": {},
   "source": [
    "Once we have created the dds object, we will apply the `deseq2` method to it. This method normalises the data, estimates the dispersion and calculates the log fold change (LFC) based on the design factor."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cff7bf9c-eaa1-4ca3-8855-2ee5e2703b7a",
   "metadata": {},
   "outputs": [],
   "source": [
    "dds.deseq2()\n",
    "dds"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "60487de3-8d0c-430e-b0b1-9d1b012e4ecf",
   "metadata": {},
   "source": [
    "Note that new elements have been added to the AnnData object, included LFC and normed counts, for example. Let's have a look at the normed counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "49ee6a17-997a-46f3-b660-0f8754271630",
   "metadata": {},
   "outputs": [],
   "source": [
    "# View the normed counts\n",
    "dds.layers['normed_counts']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b827ae6f-573b-4e78-805d-0ac481c547cf",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-success\">\n",
    "<b>Learning outcome</b>\n",
    "    \n",
    "- You should now know how to use PyDESeq2 to normalise your data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b3d5ace2-b160-4b50-a85d-9d7e60613d91",
   "metadata": {},
   "source": [
    "# PCA Plot\n",
    "\n",
    "In our experiment, we have 12 different samples (three replicates of 3 hr schistosomulum, four replicates of 24 hr schistosomulum, four replicates of cercarium and one replicate of platyhelminth adult). For each of these samples, we have gene counts for thousands of genes. It would be useful at this point to have an overview of the data. We expect the replicates within each stage to be very similar between them, but to be different from the other stages.  Principal Component Analysis (PCA) helps us do this.  You studied PCA in detail during DExB2. Look at the week 7, class 14 if you need a refresher. This [video](https://www.youtube.com/watch?v=5vgP05YpKdE) provides a simple overview of the concept. \n",
    "\n",
    "In our experiment, PCA will look at all the normalised gene counts and construct groups of genes (the \"components\") that describe as much of the variation between samples as possible. Plotting the first 2 components identified in the analysis can therefore be a useful way to visualise the effect of experimental covariates as well as batch effects.\n",
    "\n",
    "We are going to use PCA to have a look at our data. We will perform the PCA and create the plot with the python library [scanpy](https://scanpy.readthedocs.io/en/stable/)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "780979b6-8900-4c05-a5b0-6f8e046dc165",
   "metadata": {},
   "outputs": [],
   "source": [
    "! pip install --quiet scanpy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e9b96f3e-572f-4c79-9edc-75f4e04f1e89",
   "metadata": {},
   "outputs": [],
   "source": [
    "import scanpy as sc\n",
    "\n",
    "# Use scanpy to plot the first 2 components\n",
    "sc.tl.pca(dds)\n",
    "sc.pl.pca(dds, size=200, color=\"stage\", annotate_var_explained=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c44e965b-587e-4bf0-8d7b-b9717fda29a5",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-warning\">\n",
    "\n",
    "Discuss in your group:\n",
    "- Is there a separation of the different developmental stages in the PCA plot?\n",
    "- How well do the replicates for each stage cluster together?\n",
    "- How much variance is explained by the first 2 principal components?\n",
    "- What is PC1 separating?\n",
    "- What is PC2 separating?"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5cf43f45-0a20-4a4c-8c30-407e75fd2c04",
   "metadata": {},
   "source": [
    "You can add some notes from your discussion in this text cell\n",
    "\n",
    "\\\n",
    "\n",
    "\\\n",
    "\n",
    "\\"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d0d76740-48fe-49ce-bbf9-e9c907eb546a",
   "metadata": {},
   "source": [
    "We can now look at which genes are contributing to each component, and compare how much each gene contributes. This is called \"loadings\"."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2cfd9861-5791-428c-af53-0d4bb2c79255",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.pca_loadings(dds, components = '1,2')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9ba1cf74-6037-4324-b6f6-398b5164b2ea",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-warning\">\n",
    "\n",
    "Discuss in your group:\n",
    "\n",
    "- Which 3 genes contribute most to PC1?\n",
    "- Which 2 genes contribute most to PC2?"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "be1d3774-2041-4ee0-a6d2-8acf67d9b312",
   "metadata": {},
   "source": [
    "You can add some notes from your discussion in this text cell\n",
    "\n",
    "\\\n",
    "\n",
    "\\\n",
    "\n",
    "\\"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5be6f38f-fca7-4308-85eb-3de509d6f510",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-success\">\n",
    "<b>Learning outcomes</b>\n",
    "    \n",
    "You should now know...\n",
    "    \n",
    "- Why we perform PCA to visualise data\n",
    "- How to interpret a PCA plot\n",
    "- How to use scanpy to perform PCA and extract the loadings for each component"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bc2dc406-e2e2-442c-b008-9fdefe1d8237",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
